pkg/packetfilter/nftables/nftables.go

```diff
@@ -0,0 +1,511 @@
+		packetfilter.ChainTypeNAT:    knftables.NATType,
+	}
+
+	iphookChainHookToNftablesHook = map[packetfilter.ChainHook]knftables.BaseChainHook{
+		packetfilter.ChainHookPrerouting:  knftables.PreroutingHook,
+		packetfilter.ChainHookInput:       knftables.InputHook,
+		packetfilter.ChainHookForward:     knftables.ForwardHook,
+		packetfilter.ChainHookOutput:      knftables.OutputHook,
+		packetfilter.ChainHookPostrouting: knftables.PostroutingHook,
+	}
+
+	iphookChainTypeToNftablesBasicPriority = map[packetfilter.ChainType]knftables.BaseChainPriority{
+		packetfilter.ChainTypeFilter: knftables.FilterPriority,
+		packetfilter.ChainTypeRoute:  knftables.ManglePriority,
+		packetfilter.ChainTypeNAT:    knftables.DNATPriority,
+	}
+
+	ruleActionToStr = map[packetfilter.RuleAction][]string{
+		packetfilter.RuleActionAccept: {"accept"},
+		packetfilter.RuleActionMss:    {"tcp", "option", "maxseg"},
+		packetfilter.RuleActionMark:   {"meta", "mark"},
+		packetfilter.RuleActionSNAT:   {"snat"},
+		packetfilter.RuleActionDNAT:   {"dnat"},
+		packetfilter.RuleActionJump:   {"jump"},
+	}
+
+	logger = log.Logger{Logger: logf.Log.WithName("NFTables")}
+)
+
+type RuleSpec []string
+
+func (r RuleSpec) String() string {
+	return strings.Join(r, " ")
+}
+
+type packetFilter struct {
+	nftables knftables.Interface
+}
+
+func New() (packetfilter.Driver, error) {
+	nft, err := knftables.New(knftables.IPv4Family, submarinerTable)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating knftables")
+	}
+
+	return &packetFilter{
+		nftables: nft,
+	}, nil
+}
+
+func (p *packetFilter) ChainExists(_ packetfilter.TableType, chain string) (bool, error) {
+	return p.chainExists(chain)
+}
+
+func (p *packetFilter) chainExists(chain string) (bool, error) {
+	chainsList, err := p.nftables.List(context.TODO(), "chains")
+	if err != nil && !knftables.IsNotFound(err) {
+		return false, errors.Wrap(err, "error listing chains")
+	}
+
+	return slices.Contains(chainsList, chain), nil
+}
+
+func (p *packetFilter) AppendUnique(_ packetfilter.TableType, chain string, rule *packetfilter.Rule) error {
+	return p.insertRuleAtPosition(chain, rule, -1)
+}
+
+func (p *packetFilter) CreateIPHookChainIfNotExists(chain *packetfilter.ChainIPHook) error {
+	exists, err := p.chainExists(chain.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if chain exists")
+	}
+
+	if exists {
+		return nil
+	}
+
+	tx := p.nftables.NewTransaction()
+	tx.Add(&knftables.Table{
+		Comment: ptr.To("rules for submariner"),
+	})
+
+	chainPriority := iphookChainTypeToNftablesBasicPriority[chain.Type]
+	if chain.Priority == packetfilter.ChainPriorityFirst {
+		chainPriority += "-10"
+	}
+
+	nftChain := &knftables.Chain{
+		Name:     chain.Name,
+		Type:     ptr.To(iphookChainTypeToNftablesType[chain.Type]),
+		Hook:     ptr.To(iphookChainHookToNftablesHook[chain.Hook]),
+		Priority: ptr.To(chainPriority),
+	}
+	tx.Add(nftChain)
+	err = p.nftables.Run(context.TODO(), tx)
```

Comment:
We can check for `IsAlreadyExists` so we don't need to separately check `chainExists` above.
```suggestion
	err = p.nftables.Run(context.TODO(), tx)
	if knftables.IsAlreadyExists(err) {
		return nil
	}
```